### PR TITLE
[keyboard] Add Torn keyboard

### PR DESCRIPTION
**Pull request by rtitmuss**
## Description

Added support for the Torn, a split through hole keyboard.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/torn/keymaps/via/torn.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tzarc) [inline keyboards/torn/keymaps/via/torn.json:1]: Please remove this file -- via json's go into their repository, not QMK.

File: keyboards/torn/torn.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+) { \
+	{ k00, k01, k02, k03, k04, k05,   k06, k07, k08, k09, k0a, k0b }, \
+	{ k10, k11, k12, k13, k14, k15,   k16, k17, k18, k19, k1a, k1b }, \
+	{ k20, k21, k22, k23, k24, k25,   k26, k27, k28, k29, k2a, k2b }, \
+	{ ___, ___, k32, k33, k34, k35,   k36, k37, k38, k39, ___, ___ } \
+}
+// clang-format on
+
+bool torn_oled_process_record_kb(uint16_t keycode, keyrecord_t *record);
+
```
Comment (tzarc) [inline keyboards/torn/torn.h:39]: Unused function.
```suggestion
```

**Review by tzarc — CHANGES_REQUESTED** [review]

File: keyboards/torn/readme.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# TORN// Split Through Hole
+
+![torn](https://raw.githubusercontent.com/rtitmuss/torn/master/doc/img/torn.jpg)
+
+The split keyboard kit made by through hole components only.
+
+The design has been inspired by the Corne, Plaid and Discipline keyboards.
+
+* Keyboard Maintainer: [rtitmuss](https://github.com/rtitmuss)
+* Hardware Supported: TORN, atmega32a
```
Comment (rtitmuss) [inline keyboards/torn/readme.md:10]:
```suggestion
* Hardware Supported: TORN, atmega32p
```

File: keyboards/torn/matrix.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+#include "print.h"
+
+#define SPLIT_MATRIX_COLS (MATRIX_COLS / 2)
+#define SECONDARY_ROW_OFFSET (MATRIX_ROWS / 2);
+
+static const pin_t    row_pins[MATRIX_ROWS]                 = MATRIX_ROW_PINS;
+static const pin_t    col_pins[SPLIT_MATRIX_COLS]           = MATRIX_COL_PINS;
+static const pin_t    secondary_row_pins[MATRIX_ROWS]       = SECONDARY_ROW_PINS;
+static const uint16_t secondary_col_pins[SPLIT_MATRIX_COLS] = SECONDARY_COL_PINS;
```
[the remark this answers is not in the document]
Comment (rtitmuss) [reply keyboards/torn/matrix.c]: The secondary rows and cols on an i/o expander. I have added a new typedef to document that these are not the same as pin_t.

File: keyboards/torn/matrix.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+#include QMK_KEYBOARD_H
+#include "mcp23018.h"
+
+#include "print.h"
+
+#define SPLIT_MATRIX_COLS (MATRIX_COLS / 2)
+#define SECONDARY_ROW_OFFSET (MATRIX_ROWS / 2)
+
+typedef uint16_t mcp23081_pin_t;
```
Comment (noroadsleft) [inline keyboards/torn/matrix.c:27]:
```suggestion
typedef uint16_t mcp23018_pin_t;
```

**Review by noroadsleft — COMMENTED** [review]
Shouldn't these be...

File: keyboards/torn/matrix.c [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include "mcp23018.h"
+
+#include "print.h"
+
```
Comment (rtitmuss) [inline keyboards/torn/matrix.c:23]:
```suggestion
```

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

File: keyboards/torn/rules.mk [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ENCODER_ENABLE = yes        # Enable rotary encoder
+OLED_DRIVER_ENABLE = yes
+CUSTOM_MATRIX = lite
+
+SRC += \
+    i2c_master.c \
+    matrix.c \
+    mcp23018.c \
+    torn_encoder.c \
+    torn_oled.c
```
Comment (drashna) [inline keyboards/torn/rules.mk:26]:
```suggestion
SRC += matrix.c \
    torn_encoder.c \
    torn_oled.c
QUANTUM_LIB_SRC += i2c_master.c \
    mcp23018.c \
```

File: keyboards/torn/torn_oled.c [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        0x0c, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x0f, 0x0f, 0x0c, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+        0x03, 0x07, 0x0e, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+     };
+
+    oled_write_raw_P(torn_logo, sizeof(torn_logo));
+}
+// clang-format on
+
+void oled_task_user(void) { render_logo(); }
```
Comment (drashna) [inline keyboards/torn/torn_oled.c:62]:
```suggestion
__attribute__((weak))
void oled_task_user(void) { 
   static const char PROGMEM torn_logo[] = {
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf8, 0xf8, 0x98, 0x98, 0x98, 0x98, 0x98, 0x18, 0x18, 0x98,
        0x98, 0x98, 0x98, 0x98, 0xf8, 0xf8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x80, 0xc0, 0x60, 0x30, 0x30, 0x18, 0x98, 0x98, 0x98, 0x98, 0x98, 0x98, 0x98,
        0x18, 0x30, 0x30, 0x60, 0xc0, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0xf8, 0xf8, 0x18, 0x18, 0x98, 0x98, 0x98, 0x98, 0x98, 0x98, 0x18, 0x38,
        0x70, 0xe0, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xf8, 0xf8, 0x38, 0x70, 0xe0, 0xc0, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xf8, 0xf8, 0x18, 0x18, 0xf8, 0xf8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x01, 0x01, 0xff, 0xff, 0x00, 0x00, 0xff,
        0xff, 0x01, 0x01, 0x01, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xf8, 0xfe, 0x07, 0x01, 0xf0, 0xfc, 0x0e, 0x03, 0x03, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
        0x03, 0x03, 0x0e, 0xfc, 0xf0, 0x01, 0x07, 0xfe, 0xfc, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0xc1, 0xc1, 0xc1, 0x63, 0x3f, 0x1c,
        0xc0, 0xe1, 0x7f, 0x1e, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xff, 0xff, 0x00, 0x00, 0xf8, 0xf1, 0xe3, 0xc7, 0x8e, 0x1c, 0x38, 0x70, 0xe0, 0xc0, 0x80, 0x00,
        0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0xff,
        0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x0f, 0x3f, 0x70, 0xc0, 0x87, 0x1f, 0x38, 0x60, 0x60, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0,
        0x60, 0x60, 0x38, 0x1f, 0x87, 0xc0, 0xf0, 0x7f, 0x1f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0xfe, 0xfe, 0x0e, 0x3c, 0x78, 0xe2, 0xc7, 0x1f,
        0x39, 0xe0, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0x00, 0x01, 0x03, 0x07, 0x0e, 0x1c, 0x38, 0x71, 0xe3, 0xc7,
        0x8f, 0x1f, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0f, 0x0f, 0x0c, 0x0c, 0x0f,
        0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x01, 0x03, 0x06, 0x06, 0x0c, 0x0c, 0x0c, 0x0c, 0x0c, 0x0c, 0x0c, 0x0c,
        0x0c, 0x06, 0x06, 0x03, 0x03, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x0f, 0x0f, 0x0c, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x01, 0x07, 0x0f,
        0x0c, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x0f, 0x0f, 0x0c, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x03, 0x07, 0x0e, 0x0c, 0x0f, 0x0f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
     };

    oled_write_raw_P(torn_logo, sizeof(torn_logo));
}
// clang-format on

```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/torn/rules.mk (hunk 7 above)
Comment (rtitmuss) [reply keyboards/torn/rules.mk:26]:
```suggestion
SRC += \
    matrix.c \
    mcp23018.c \
    torn_encoder.c \
    torn_oled.c
QUANTUM_LIB_SRC += i2c_master.c
```

Comment (rtitmuss) [reply keyboards/torn/rules.mk:26]: Thanks, mcp23018.c is part of the keyboard source, but I've made the change to use QUANTUM_LIB_SRC for i2c_master.c.

**Review by drashna — APPROVED** [review]

File: keyboards/torn/rules.mk [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = no        # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+ENCODER_ENABLE = yes        # Enable rotary encoder
+OLED_DRIVER_ENABLE = yes
+CUSTOM_MATRIX = lite
```
Comment (fauxpark) [inline keyboards/torn/rules.mk:19]:
```suggestion
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
ENCODER_ENABLE = yes        # Enable rotary encoder
OLED_DRIVER_ENABLE = yes
CUSTOM_MATRIX = lite
```

File: keyboards/torn/torn.c [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    matrix_scan_user();
+    secondary_encoder_read();
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    // put your per-action keyboard code here
+    // runs for every action, just before processing by the firmware
+    return process_record_user(keycode, record);
+}
+
```
Comment (fauxpark) [inline keyboards/torn/torn.c:49]:
```suggestion
```

File: keyboards/torn/keymaps/via/rules.mk [hunk 11]
```diff
@@ -0,0 +1,5 @@
+VIA_ENABLE = yes
+LTO_ENABLE = yes
+
+MOUSEKEY_ENABLE = no
+EXTRAKEY_ENABLE = no
```
Comment (fauxpark) [inline keyboards/torn/keymaps/via/rules.mk:5]: These should work now.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/torn/keymaps/via/rules.mk (hunk 11 above)
Comment (rtitmuss) [inline keyboards/torn/keymaps/via/rules.mk:5]:
```suggestion

MOUSEKEY_ENABLE = no
EXTRAKEY_ENABLE = no
```

Comment (rtitmuss) [reply keyboards/torn/keymaps/via/rules.mk]: I've removed these

**Pull request merged**